### DISCUSSION TITLE
Clean up directories (possibly with extra non-installed files) on uninstall (but not major upgrade)

### DISCUSSION THREAD

**Discussion by mr-intj** [wixtoolset/issues#6414] [Questions]
Our app keeps some system-wide configuration files under `C:\ProgramData\<OurAppName>`, so the install creates that directory and a few sub-directories:

```
  C:\ProgramData
     +- <OurAppName>
        +- Foo1
        |  +- Foo2
        +- Bar1
        |  +- Bar2
        +- Baz1
           +- Baz2
```

- The app's configuration files, plus _foo_ and _bar_ and _baz_ files, etc are created in the sub-directories at runtime. An uninstall never removes any of these directories, of course, because they contain files that weren't installed.

- When a user performs a _major_ upgrade, though, we'd like to **_not_** remove the configuration information or other files

- But when a user is doing a stand-alone uninstall (vs. nested as part of an upgrade), we would like to remove `.../OurAppName/*` (regardless of whether it was installed or not)

I've found a number of posts about this on stackoverflow, etc, but no one seems to have a good solution that doesn't involve a custom action (which I'm led to believe are **evil**).

I've been able to determine that the condition I need is: `(NOT UPGRADINGPRODUCTCODE) AND (REMOVE="ALL")`

I initially bundled a bunch of `<RemoveFolder ...>` and `<RemoveFile ... Name="*.*" ...>` elements in the component containing the `<CreateFolder ...>` elements and the registry setting as the keypath. Of course, that just unconditionally removes _everything_ on uninstall; even during a major upgrade.

To apply the above condition to these operations, I was going to create a separate component with just the RemoveFolder and RemoveFile elements, and condition the component on the expression above. Alas, several posts pointed out that since this new "component" will never be installed, the removals won't be executed during an uninstall either.

Is there some way to add entries to the `RemoveFile` table directly, based upon some condition?

What is the current consensus on how to do this?

Thanks.

**Comment by BMurri** [comment]
If the component Id (a guid) is unchanged for the component that contains the CreateFolder, it won't be removed during upgrades, but will be removed when running uninstall outside of a major upgrade. No condition necessary.

Search for `msi component rules` for more details.

**Reply by mr-intj** [reply]
So for the following structure...:
```
C:\ProgramData
   +- Base        (this and all subfolders contain files generated at runtime)
      +- Foo
         +- Bar
```
...if I bundle the `RemoveFile` elements in the same component with the `CreateFolder` elements...:


```
<Component ... Guid="E8B39818-B77B-4D2F-945E-7CDBE579E5EA">
    <CreateFolder Directory="dir_Base" />
    <CreateFolder Directory="dir_Base_Foo" />
    <CreateFolder Directory="dir_Base_Foo_Bar" />

    <RemoveFile Id="..." Directory="dir_Base" Name="*.*" On="uninstall" />
    <RemoveFile Id="..." Directory="dir_Base" Name="*.*" On="uninstall" />
    <RemoveFile Id="..." Directory="dir_Base" Name="*.*" On="uninstall" />

    <RegistryValue ... KeyPath="yes" ... />
</Component>
```

...then on a stand-alone uninstall (and not on an install or upgrade), the Windows installer will first remove the files from the directories, and will then remove the (now empty) directories themselves?

**Reply by BMurri** [reply]
Yes
